fix nil function invocation in client-go/tools/leaderelection

hbagdi opened the pull request:
It should be possible for the user of leaderelection API to not provide
callbacks for stopped and started leading conditions. Currnetly, the
code results in fatal error and exits.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

hbagdi commented:
I signed it

neolit123 reviewed (commented):
thank you for the PR, i think this make sense, but needs approval by the maintainers.
the 1.15 code freeze will be lifted after 11th of june.
try pinging #sig-api-machinery on slack if needed after that.

/ok-to-test
/priority backlog

fedebongio commented:
/assign @cheftako @roycaihw

cheftako commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go:
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
 		runtime.HandleCrash()
-		le.config.Callbacks.OnStoppedLeading()
+		if le.config.Callbacks.OnStoppedLeading != nil {
+			go le.config.Callbacks.OnStoppedLeading()
```
Probably not a good idea to switch this to a new goroutine without serious investigation. Our general behavior in things like the KCM to losing leader election is to kill our own process. Moving this to a goroutine means that the callback is now racing to act prior to the process being killed.

hbagdi replied:
That makes sense. I re-uped the commit and removed the goroutine.

cheftako reviewed (commented):
/lgtm

roycaihw commented:
looks good

/assign @mikedanese
for approval

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go:
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	go le.config.Callbacks.OnStartedLeading(ctx)
+	if le.config.Callbacks.OnStartedLeading != nil {
```
The other one is fine, but if this is nil, we should probably panic. There's no point to this client if you don't do anything when you get leadership.

hbagdi replied:
A client could potentially not actively want to be notified of it being a leader and but choose to check the leader status using `IsLeader()` function whenever it is actually going to take up any action.

mikedanese replied:
You can do this by providing an empty OnStartedLeading callback. I am hesitant to further facilitate this behavior.

cheftako replied:
Mike has a good point. That having been said I personally would rather an explicit panic which explained the issue (eg. "Leader elect called without a proper started leading method") than a fatal error about referencing invalid memory.

hbagdi replied:
>  You can do this by providing an empty OnStartedLeading callback. I am hesitant to further facilitate this behavior.

My take is that zero values in Go shouldn't panic in such a case.

Please let me know the exact change that should be made to move this forward. You guys are more familiar with the API style and end-user expectations here so I'll make a change that works for everyone.

mikedanese commented on staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go:
```diff
@@ -190,14 +190,18 @@ type LeaderElector struct {
 func (le *LeaderElector) Run(ctx context.Context) {
 	defer func() {
 		runtime.HandleCrash()
-		le.config.Callbacks.OnStoppedLeading()
```
Sorry I misread this originally. This is intended to crash.

mikedanese commented:
Can you validate that OnStart and OnFinish callbacks are non-nil in NewLeaderElector, else return an error?

hbagdi committed:
fix: do not allow nil Callbacks functions

If user specifies a nil callback, then error out
rather than crashing at runtime due to a nil reference.

hbagdi commented:
@mikedanese updated. Please re-review.

mikedanese commented:
/lgtm
/approve

hbagdi commented:
How do we fix the failing CI? 🤔

mikedanese commented:
/retest
